### PR TITLE
Added file protocol layer.

### PR DESCRIPTION
**Pull request by katmsft**

### CODE REVIEW

File: sdk/storage/CMakeLists.txt [hunk 1]
```diff
@@ -138,9 +138,25 @@ set (AZURE_STORAGE_DATALAKE_SOURCE
 add_library(azure-storage-file-datalake ${AZURE_STORAGE_DATALAKE_HEADER} ${AZURE_STORAGE_DATALAKE_SOURCE})
 target_link_libraries(azure-storage-file-datalake azure-storage-blob)
 
+set (AZURE_STORAGE_SHARES_HEADER
```
Comment (JinmingHu-MSFT) [inline sdk/storage/CMakeLists.txt:141]: `AZURE_STORAGE_FILE_SHARES_HEADER`?

File: sdk/storage/CMakeLists.txt [hunk 2]
```diff
@@ -138,9 +138,25 @@ set (AZURE_STORAGE_DATALAKE_SOURCE
...
+    inc/shares/shares.hpp
```
Comment (JinmingHu-MSFT) [inline sdk/storage/CMakeLists.txt:142]: What is the directory hierarchy in other SDK?

File: sdk/storage/inc/shares/service_client.hpp [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+     * @brief Gets the file share service's primary uri endpoint.
+     *
+     * @return The file share service's primary uri endpoint.
+     */
+    std::string GetUri() const { return m_serviceUri.ToString(); }
+
+    /**
+     * @brief List the shares from the service.
+     * @param options Optional parameters to list the shares.
+     * @return Azure::Core::Response<ListSharesResult>
```
Comment (JinmingHu-MSFT) [inline sdk/storage/inc/shares/service_client.hpp:73]: still, this doesn't count as documents

File: sdk/storage/inc/shares/share_options.hpp [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+     * @brief Specifies the maximum number of entries to return. If the request does not specify
+     * maxresults, or specifies a value greater than 5,000, the server will return up to 5,000
+     * items.
+     */
+    Azure::Core::Nullable<int32_t> MaxResults;
+
+    /**
+     * @brief Include this parameter to specify one or more datasets to include in the response.
+     */
+    Azure::Core::Nullable<std::vector<ListSharesIncludeType>> ListSharesInclude;
```
Comment (JinmingHu-MSFT) [inline sdk/storage/inc/shares/share_options.hpp:57]: enum?

File: sdk/storage/inc/shares/shares.hpp [hunk 5]
```diff
@@ -0,0 +1,4 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+#pragma once
```
Comment (JinmingHu-MSFT) [inline sdk/storage/inc/shares/shares.hpp:4]: include share_service.hpp?

File: sdk/storage/sample/CMakeLists.txt [hunk 6]
```diff
@@ -9,6 +9,6 @@ add_executable (
     samples_common.hpp
     blob_getting_started.cpp
     datalake_getting_started.cpp
-)
+    share_getting_started.cpp)
```
Comment (JinmingHu-MSFT) [inline sdk/storage/sample/CMakeLists.txt:12]: I think this should be called `file_share` or something

File: sdk/storage/sample/share_getting_started.cpp [hunk 7]
```diff
@@ -0,0 +1,9 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+#include "datalake/datalake.hpp"
+#include "samples_common.hpp"
+#include <iostream>
+
+SAMPLE(ShareGettingStarted, ShareGettingStarted)
+void ShareGettingStarted() {}
```
Comment (JinmingHu-MSFT) [inline sdk/storage/sample/share_getting_started.cpp:9]: FileShare?

File: sdk/storage/inc/shares/protocol/share_rest_client.hpp [hunk 8]
```diff
@@ -0,0 +1,6898 @@
...
+    }
+    if (listSharesIncludeType == "deleted")
+    {
+      return ListSharesIncludeType::Deleted;
+    }
+    throw std::runtime_error(
+        "Cannot convert " + listSharesIncludeType + " to ListSharesIncludeType");
+  }
+
+  struct ShareFileHttpHeaders
```
Comment (JinmingHu-MSFT) [inline sdk/storage/inc/shares/protocol/share_rest_client.hpp:162]: FileShare?

File: sdk/storage/inc/shares/protocol/share_rest_client.hpp [hunk 9]
```diff
@@ -0,0 +1,6898 @@
...
+    std::string AllowedHeaders; // The request headers that the origin domain may specify on the
+                                // CORS request.
+    std::string ExposedHeaders; // The response headers that may be sent in the response to the CORS
+                                // request and exposed by the browser to the request issuer.
+    int32_t MaxAgeInSeconds; // The maximum amount time that a browser should cache the preflight
+                             // OPTIONS request.
+  };
+
+  // Error codes returned by the service
+  enum class StorageErrorCode
```
Comment (JinmingHu-MSFT) [inline sdk/storage/inc/shares/protocol/share_rest_client.hpp:300]: Is this used anywhere?

File: sdk/storage/inc/shares/protocol/share_rest_client.hpp [hunk 10]
```diff
@@ -0,0 +1,6898 @@
...
+  struct FileCreateResponse
+  {
+    std::string ETag;
+    std::string LastModified;
+    bool IsServerEncrypted = bool();
+    std::string FilePermissionKey;
+    std::string FileAttributes;
+    std::string FileCreationTime;
+    std::string FileLastWriteTime;
+    std::string FileChangeTime;
```
Comment (JinmingHu-MSFT) [inline sdk/storage/inc/shares/protocol/share_rest_client.hpp:1392]: Do we need `File` prefix for these?

File: sdk/storage/inc/shares/protocol/share_rest_client.hpp [hunk 11]
```diff
@@ -0,0 +1,6898 @@
...
+    class Service {
+    public:
+      struct SetPropertiesOptions
+      {
+        StorageServiceProperties ServiceProperties; // The StorageService properties.
+        Azure::Core::Nullable<int32_t>
+            Timeout; // The timeout parameter is expressed in seconds. For more information, see <a
+                     // href="https://docs.microsoft.com/en-us/rest/api/storageservices/Setting-Timeouts-for-File-Service-Operations?redirectedfrom=MSDN">Setting
+                     // Timeouts for File Service Operations.</a>
+        std::string ApiVersionParameter
```
Comment (JinmingHu-MSFT) [inline sdk/storage/inc/shares/protocol/share_rest_client.hpp:1568]: why is this in options? Do we allow users to specify API version?

File: sdk/storage/inc/shares/protocol/share_rest_client.hpp [hunk 12]
```diff
@@ -0,0 +1,6898 @@
...
+        request.AddHeader(Details::c_HeaderVersion, listSharesSegmentOptions.ApiVersionParameter);
+        return ListSharesSegmentParseResponse(context, pipeline.Send(context, request));
+      }
+
+    private:
+      static Azure::Core::Response<ServiceSetPropertiesResponse> SetPropertiesParseResponse(
+          Azure::Core::Context context,
+          std::unique_ptr<Azure::Core::Http::RawResponse> responsePtr)
+      {
+        /* const */ auto& response = *responsePtr;
```
Comment (JinmingHu-MSFT) [inline sdk/storage/inc/shares/protocol/share_rest_client.hpp:1707]: We don't need const here

File: sdk/storage/CMakeLists.txt (hunk 2 above)
Comment (katmsft) [reply sdk/storage/CMakeLists.txt:142]: 5 separate services has it's own inc/src folders. I believe we can change them when we pin down the final decision on folder hierachy.

File: sdk/storage/CMakeLists.txt (hunk 1 above)
Comment (katmsft) [reply sdk/storage/CMakeLists.txt:141]: But we are also calling `AZURE_STORAGE_DATALAKE_HEADER`, do you think it's necessary to change it into `AZURE_STORAGE_FILE_DATALAKE_HEADER`?

Comment (JinmingHu-MSFT) [reply sdk/storage/CMakeLists.txt:141]: I don't know, whichever you like.

i'm just not okay with `AZURE_STORAGE_SHARES_HEADER`

File: sdk/storage/inc/shares/protocol/share_rest_client.hpp (hunk 9 above)
Comment (katmsft) [reply sdk/storage/inc/shares/protocol/share_rest_client.hpp:300]: Nope, removed.

File: sdk/storage/inc/shares/protocol/share_rest_client.hpp (hunk 10 above)
Comment (katmsft) [reply sdk/storage/inc/shares/protocol/share_rest_client.hpp:1392]: Yes, to align with .NET.

Comment (JinmingHu-MSFT) [reply sdk/storage/inc/shares/protocol/share_rest_client.hpp:1392]: They have `FileSmbProperties`

File: sdk/storage/inc/shares/protocol/share_rest_client.hpp (hunk 11 above)
Comment (katmsft) [reply sdk/storage/inc/shares/protocol/share_rest_client.hpp:1568]: It's never exposed to convenience layer.

Comment (JinmingHu-MSFT) [reply sdk/storage/inc/shares/protocol/share_rest_client.hpp:1097]: I don't think it's ok. If our customers saw this, they'd think we're planning on support this feature, which we're not.

File: sdk/storage/inc/shares/shares.hpp [hunk 13]
```diff
@@ -2,3 +2,5 @@
 // SPDX-License-Identifier: MIT
 
 #pragma once
+
+#include "service_client.hpp"
```
Comment (JinmingHu-MSFT) [inline sdk/storage/inc/shares/shares.hpp:6]: new line

File: sdk/storage/CMakeLists.txt (hunk 2 above)
Comment (katmsft) [reply sdk/storage/CMakeLists.txt:142]: https://github.com/Azure/azure-sdk-for-cpp/issues/339 this issue was opened to track the change.

File: sdk/storage/inc/shares/protocol/share_rest_client.hpp (hunk 11 above)
Comment (katmsft) [reply sdk/storage/inc/shares/protocol/share_rest_client.hpp:1097]: Added this into the discussion topic list.

File: sdk/storage/inc/shares/protocol/share_rest_client.hpp (hunk 10 above)
Comment (katmsft) [reply sdk/storage/inc/shares/protocol/share_rest_client.hpp:921]: Convenience layer will not have this.

File: sdk/storage/CMakeLists.txt (hunk 1 above)
Comment (katmsft) [reply sdk/storage/CMakeLists.txt:141]: Sure. :)

**Pull request merged**